### DISCUSSION TITLE
Single input date field

### DISCUSSION THREAD

**Discussion by owensgit** [Legal-and-General/canopy#578] [Ideas] [answered]
Currently we have a **multiple** input date field:

<img width="283" alt="Screenshot 2021-10-08 at 12 16 46" src="https://user-images.githubusercontent.com/1943532/136547444-80b46f68-cde7-4087-add1-401f1fc5e902.png">

However there is a requirement to create a **single** input field:

<img width="208" alt="Screenshot 2021-10-08 at 12 17 20" src="https://user-images.githubusercontent.com/1943532/136547584-c59c1880-73f5-43ed-a8d0-4f6c698e0a86.png">

Note - the calendar icon button is not included in this proposal for now - it will eventually be added to open a date picker. I'm proposing that the date picker will be a entirely separate component that could be used to populate this new field (or others) if required.

For now, the idea is to just build the single input without the button, which encapsulates the functionality below, so that it's consistent when used across each L&G app. 

### Functionality

As discussed with @michelleallendavey.

- User types a date into a single input
- Ideally they enter a date as `DD MM YYYY` as per the placeholder, with spaces
- Placeholder is auto generated because the requirements are always the same
- `/`, `.` and `-` characters are allowed but will be stripped out and replaced with spaces when the input is blurred. e.g. `08/10/2021` will become `08 10 2021`
- Maximum of 10 characters allowed to account for all DD MM YYYY chars, plus the  `/`, `.` and `-` chars.
- If the characters can not be parsed as a valid date, it will throw an invalid date error
- We will not restrict the input to only accept numbers and `/`, `.` or `-` , instead an valid date error will be shown as above if other characters are entered

### Output

- The form model should be updated with a date string, the same as the current Date Field component - e.g. `2021-10-08`
- Custom validators can be added for min/max dates, similar to the ones we already have, or hopefully can be reused.

### Directive

```
<lg-input-field>
  Retirement date
  <input lgInput lgDateInput formControlName="date" />
  <button lg-button lgSuffix size="sm" [iconButton]="true" variant="add-on">
    <lg-icon name="calandar" />
  </button>
</lg-input-field>
```

I'd need to validate that this would work, because it's a directive being added to a control 🤔  But the idea is that it should allow us to add the above functionality, while still allowing the developer to customise it just like a regular input with the prefix/suffix buttons, and the usual hint and validation messages etc. It would also allow for adding click events to the icon, for opening up a future date picker. Again this depends on whether combining an input with a directive for overriding form control behavior is a good idea.

### Encapsulated in a component

Similar to the current date field, it could be encapsulated in a component and implemented as a custom control, with content projection slots for the label, unit and validation messages.

```
<lg-date-field-single showCalandarButton="true">
  Retirement date
  <lg-hint>For example, 12 06 1983</lg-hint>
</lg-date-field-single>
```

The input `showCalandarButton` will show/hide the calendar icon button and the component would emit an event when it is clicked, allowing the developer to open the date picker when it is build.

**Comment by michelleallendavey** [comment]
Have walked through the functionality with Owen (I've been designing the date picker / date picker input) and it all makes sense to me from a user perpective.

**Reply by aaronBery** [reply]
I'm just thinking about the validation for the inputs....are there any additional considerations for languages that don't use the numbers 0-9?

**Reply by owensgit** [reply]
My understanding is that there's no requirement to support this, but would be happy to discuss. 
The current date field doesn't, so if we decided to implement it on this new component we should also look to update the existing one too.

**Comment by owensgit** [comment, marked as answer]
After a discussion with @elenagarrone @paul-request, we decided that the best approach would be to take inspiration from the Material UI date picker, where by it can be paired with an input field using a directive:

```
<mat-form-field appearance="fill">
  <mat-label>Choose a date</mat-label>
  <input matInput [matDatepicker]="picker">
  <mat-datepicker-toggle matSuffix [for]="picker"></mat-datepicker-toggle>
  <mat-datepicker #picker></mat-datepicker>
</mat-form-field>
```

For example, in Canopy, we could have something such as:

```
<lg-input-field>
  Retirement date
  
  <!-- Input is linked to the date picker component via a directive -->
  <input lgInput [lgDatePicker]="picker" />

  <!-- Button can also be tied to a date picker in some way -->
  <button lg-button lgSuffix size="sm" [iconButton]="true" variant="add-on">
    <lg-icon name="calandar" />
  </button>

  <!-- Date picker itself can be positioned relative to the input field -->
  <lg-datepicker #picker></lg-datepicker>
</lg-input-field>
```
